src/main/java/org/sciborgs1155/robot/drivetrain/SparkDrive.java

```diff
@@ -0,0 +1,168 @@
+package org.sciborgs1155.robot.drivetrain;
+
+import static org.sciborgs1155.robot.drivetrain.DriveConstants.STARTING_POSE;
+
+import com.revrobotics.CANSparkBase.IdleMode;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.RelativeEncoder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.math.kinematics.DifferentialDriveWheelPositions;
+import monologue.Annotations.Log;
+import org.sciborgs1155.lib.FaultLogger;
+
+/** {@link DriveIO} class using CANSparkMax controllers */
+public class SparkDrive implements DriveIO {
+  private CANSparkMax frontLeftMotor;
+  private CANSparkMax rearLeftMotor;
+  private CANSparkMax frontRightMotor;
+  private CANSparkMax rearRightMotor;
+
+  private RelativeEncoder rearLeftEncoder;
+  private RelativeEncoder frontLeftEncoder;
+  private RelativeEncoder frontRightEncoder;
+  private RelativeEncoder rearRightEncoder;
+
+  /** We don't have a gyro, so this has to be used with an estimated rotation */
+  private final DifferentialDriveOdometry odometry =
```

Comment:
This can be included in the `Drive` subsystem instead. If you take a look at your sim impl as well, you will see that you also have an odometry object in it. You can extract `odometry` to Drive such that you remove this duplication by feeding odometry update measurements into the object in Drive

Reply:
This can also be applied for many of the other storage and utility objects below. If you would like a reference, see how we use the odometry and kinematics classes between subsystems and their implementations in Crescendo (in drive)

Reply:
Aight, had to do some freaky stuff for sim interface(namely the 'update' method)